wfe2/verify.go

```diff
@@ -5,246 +5,540 @@ import (
+	// `jose.parseSigned` since it will strip away these headers.
+	var unprotected struct {
+		Header map[string]string
+	}
+	if err := json.Unmarshal(bodyBytes, &unprotected); err != nil {
+		wfe.stats.Inc("Errors.JWSParseError", 1)
+		logEvent.AddError("Parse error reading JWS from POST body")
+		return nil, "", probs.Malformed("Parse error reading JWS")
+	}
+
+	// ACME v2 never uses values from the unprotected JWS header. Reject JWS that
+	// include unprotected headers.
+	if unprotected.Header != nil {
+		wfe.stats.Inc("Errors.JWSUnprotectedHeaders", 1)
+		errMsg := "Unprotected headers included in JWS"
+		logEvent.AddError(errMsg)
+		return nil, "", probs.Malformed(errMsg)
+	}
+
+	// Parse the JWS using go-jose and enforce that the expected one non-empty
+	// signature is present in the parsed JWS.
 	body := string(bodyBytes)
+	parsedJWS, err := jose.ParseSigned(body)
+	if err != nil {
+		wfe.stats.Inc("Errors.JWSParseError", 1)
+		logEvent.AddError("Parse error reading JWS from POST body")
+		return nil, "", probs.Malformed("Parse error reading JWS")
+	}
+	if len(parsedJWS.Signatures) > 1 {
+		wfe.stats.Inc("Errors.TooManySignaturesInJWS", 1)
+		logEvent.AddError("Too many signatures in POST body JWS")
+		return nil, "", probs.Malformed("Too many signatures in POST body")
+	}
+	if len(parsedJWS.Signatures) == 0 {
+		wfe.stats.Inc("Errors.NoSignaturesInJWS", 1)
+		logEvent.AddError("POST JWS not signed")
+		return nil, "", probs.Malformed("POST JWS not signed")
+	}
+	if len(parsedJWS.Signatures) == 1 && len(parsedJWS.Signatures[0].Signature) == 0 {
+		wfe.stats.Inc("Errors.EmptySignatureInJWS", 1)
+		logEvent.AddError("POST JWS not signed")
+		return nil, "", probs.Malformed("POST JWS not signed")
+	}
 
-	// Verify JWS
-	// NOTE: It might seem insecure for the WFE to be trusted to verify
-	// client requests, i.e., that the verification should be done at the
-	// RA.  However the WFE is the RA's only view of the outside world
-	// *anyway*, so it could always lie about what key was used by faking
-	// the signature itself.
-	submittedKey, parsedJws, err := wfe.extractJWSKey(body)
+	return parsedJWS, body, nil
+}
+
+// extractJWK extracts a JWK from a provided JWS or returns a problem. It
+// expects that the JWS is using the embedded JWK style of authentication and
+// does not contain an embedded Key ID. Callers should have acquired the
+// provided JWS from parseJWS to ensure it has the correct number of signatures
+// present. The provided logEvent is mutated to add any observed errors.
+func (wfe *WebFrontEndImpl) extractJWK(
+	jws *jose.JSONWebSignature,
+	logEvent *requestEvent) (*jose.JSONWebKey, *probs.ProblemDetails) {
+	// extractJWK expects the request to be using an embedded JWK auth type and
+	// to not contain the mutually exclusive KeyID.
+	if prob := wfe.enforceJWSAuthType(jws, embeddedJWK); prob != nil {
+		logEvent.AddError("JWS auth type was not expected embeddedJWK auth")
```

Comment:
Similar here; let's not duplicate log messages. And if you remove the duplication, you also get to remove a method parameter (`logEvent`).